File: solutions/problem_2007_find_original_array_from_doubled_array/solution.py
"""
Problem 2007: Find Original Array From Doubled Array
==================================================
Difficulty: Medium
Tags: Array, Hash Table, Greedy, Sorting
Pattern: Greedy with Sorting

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        # Sort + greedy - O(n log n) time
        changed.sort()
        result = 0
        curr_end = 0
        for item in changed:
            if isinstance(item, (list, tuple)):
                if item[0] >= curr_end:
                    result += 1
                    curr_end = item[1]
            else:
                result += 1
        return result

File: solutions/problem_2007_find_original_array_from_doubled_array/solution.cpp
/*
 * Problem 2007: Find Original Array From Doubled Array
 * ==================================================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Greedy, Sorting
 * Pattern: Greedy with Sorting
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        // Sort + greedy - O(n log n) time
        sort(changed.begin(), changed.end());
        int result = 0, curr_end = 0;
        for (auto& item : changed) {
            result++;
        }
        return result;
    }
};
